plugins/browser-plugin-youtube-tracking/src/index.ts

```diff
@@ -0,0 +1,231 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import { UrlParameters, trackingOptionsParser } from './helperFunctions';
+import { YTError, YTPlayerEvent, YTState, YTStateEvent } from './constants';
+import { EventData, MediaTrackingOptions, TrackedPlayer, TrackingOptions } from './types';
+import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
+import { buildSelfDescribingEvent, CommonEventProperties, Logger, SelfDescribingJson } from '@snowplow/tracker-core';
+import { SnowplowEvent } from './snowplowEvents';
+import { MediaPlayerEvent } from './contexts';
+import { buildYouTubeEvent } from './buildYouTubeEvent';
+import { YTEvent } from './youtubeEvents';
+
+declare global {
+  interface Window {
+    onYouTubeIframeAPIReady?: Function;
+  }
+}
+
+let LOG: Logger;
+export const _trackers: Record<string, BrowserTracker> = {};
+let tracked_players: Record<string, TrackedPlayer> = {};
+let iframeApiHasLoaded = false;
+let trackingQueue: Array<TrackingOptions> = [];
+
+export function YouTubeTrackingPlugin(): BrowserPlugin {
+  return {
+    activateBrowserPlugin: (tracker: BrowserTracker) => {
+      _trackers[tracker.id] = tracker;
+    },
+    logger: (logger) => {
+      LOG = logger;
+    },
+  };
+}
+
+function trackEvent(
+  event: SelfDescribingJson<MediaPlayerEvent> & CommonEventProperties,
+  trackers: Array<string> = Object.keys(_trackers)
+): void {
+  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
+    t.core.track(buildSelfDescribingEvent({ event }), event.context, event.timestamp);
+  });
+}
+
+function enableYouTubeIframeAPI() {
+  const tag: HTMLScriptElement = document.createElement('script');
+  tag.src = 'https://www.youtube.com/iframe_api';
+  const firstScriptTag = document.getElementsByTagName('script')[0];
+  firstScriptTag.parentNode!.insertBefore(tag, firstScriptTag);
+}
+
+export function enableYouTubeTracking(args: { id: string; options?: MediaTrackingOptions }) {
+  const conf: TrackingOptions = trackingOptionsParser(args.id, args.options);
+  const el: HTMLIFrameElement = document.getElementById(args.id) as HTMLIFrameElement;
+  if (!el) {
+    LOG.error('Cannot find YouTube IFrame element');
+    return;
+  }
+  conf.urlParameters = new UrlParameters(el.src);
+  el.src = conf.urlParameters.getUrl();
+  enableYouTubeIframeAPI();
+
+  // Function to run when the API is loaded
+  window.onYouTubeIframeAPIReady = () => {
+    iframeApiHasLoaded = true;
+    trackingQueue.forEach((conf) => {
+      addListeners(conf);
+    });
+  };
+
+  // If the API hasn't loaded, it will be added to the queue
+
+  if (iframeApiHasLoaded) {
+    addListeners(conf);
+  } else {
+    trackingQueue.push(conf);
+  }
+}
+
```

Comment:
In a similar vein to the above... This section seems important to get stuff set up... however, what is it's already loaded before the tracker loads (i.e. the script it already on the page), would `window.onYouTubeIframeAPIReady` even fire?